src/pytorch_lightning/strategies/colossalai.py

```diff
@@ -0,0 +1,294 @@
+from typing import Any, Dict, List, Optional
+
+import torch
+from lightning_utilities.core.imports import RequirementCache
+
+import pytorch_lightning as pl
+from lightning_lite.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.accelerators.cuda import CUDAAccelerator
+from pytorch_lightning.overrides.base import _LightningModuleWrapperBase, _LightningPrecisionModuleWrapperBase
+from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
+from pytorch_lightning.plugins.precision import ColossalAIPrecisionPlugin, PrecisionPlugin
+from pytorch_lightning.strategies.ddp import DDPStrategy
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+
+_COLOSSALAI_AVAILABLE = RequirementCache("colossalai")
+if _COLOSSALAI_AVAILABLE:
+    from colossalai.context import ParallelMode
+    from colossalai.core import global_context as gpc
+    from colossalai.gemini import ChunkManager, GeminiManager
+    from colossalai.logging import disable_existing_loggers, get_dist_logger
+    from colossalai.nn.optimizer import CPUAdam, HybridAdam
+    from colossalai.nn.parallel import ZeroDDP
+    from colossalai.tensor import ProcessGroup
+    from colossalai.utils import get_current_device
+    from colossalai.utils.model.colo_init_context import ColoInitContext
+    from colossalai.zero import ZeroOptimizer
+
+
+class ColossalAIStrategy(DDPStrategy):
+    """ColossalAI strategy. It only supports single optimizer which must be  `colossalai.nn.optimizer.CPUAdam`_ or
+    `colossalai.nn.optimizer.HybridAdam`_ now. You must initialize your model in ``configure_sharded_model()``.
+
+    It configures accelerator and precision, and you should not configure them when initializing ``Trainer``.
+    CUDA is essential for this strategy. Please make sure CUDA is available.
+
+    Example::
+
+        class GLUETransformer(LightningModule):
+            ...
+            def configure_sharded_model(self) -> None:
+                self.model = BertForSequenceClassification.from_pretrained('bert-base-uncased')
+            def on_load_checkpoint(self, checkpoint) -> None:
+                if not hasattr(self, 'model'):
+                    self.configure_sharded_model()
+        trainer = Trainer(..., strategy=ColossalAIStrategy())
```

Comment:
```suggestion
        class GLUETransformer(LightningModule):
            ...

            def configure_sharded_model(self) -> None:
                self.model = BertForSequenceClassification.from_pretrained("bert-base-uncased")

        trainer = Trainer(..., strategy="colossalai")
```
why `on_load_checkpoint` is required?

Reply:
When users want to load checkpoint, their models may not ready yet. So we need to help them initailize models.